datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java

```diff
@@ -346,16 +353,38 @@ private synchronized boolean stopEventThread() {
     return false;
   }
 
+  // Waiting for the event thread to die.
   private synchronized boolean waitForEventThreadToJoin() {
+    CoordinatorEventProcessor eventThread = getEventThread();
+    if (!eventThread.isAlive()) {
+      return false;
+    }
+    // Waits to acquire the Coordinator object for a maximum of _heartbeat period.
+    // The time bound waiting prevents the caller thread to not infinitely wait if
+    // the event thread is already shutdown.
+    waitForNotificationFromEventThread(_heartbeatPeriod);
     try {
-      _eventThread.join(EVENT_THREAD_LONG_JOIN_TIMEOUT);
+      _log.info("Waiting for {} milliseconds for the event thread to die.", EVENT_THREAD_LONG_JOIN_TIMEOUT);
+      eventThread.join(EVENT_THREAD_LONG_JOIN_TIMEOUT);
     } catch (InterruptedException e) {
       _log.warn("Exception caught while waiting the event thread to stop", e);
       return true;
     }
     return false;
   }
 
+  // Waits for a notification for specified duration from the event thread before acquiring the Coordinator object.
+  private void waitForNotificationFromEventThread(Duration duration) {
+    try {
```

Comment:
Is there anyway to validate if the object lock is held when we enter this method (as we are calling wait here) ?

Reply:
Validate that the wait releases the lock on the "this" object? Since its an intrinsic property of the wait function to release the lock on the object from where it is called, not sure if we need validation in code. 

In the tests, it should be validating that. I have tried reproducing the deadlock scenario in one of the tests and ensuring no deadlock would mean that the object lock is released when the zk session expiry thread is waiting.